### PR TITLE
ROX-17702: `GetByQuery` product usage

### PR DESCRIPTION
**Pull request by parametalol**
## Description

Make use of the newly implemented query by timerange.

Switched the signatures from `*types.Timestamp` to `time.Time` due to the new query interface.

Based on #7179.
Proto: #7136.

Planned optimization: make use of the pagination + sorting with `Limit(1)` to query Max values instead of computing in the backend.

## Checklist
- [x] Investigated and inspected CI test results
- [x] Unit test and regression tests added
- [ ] ~Evaluated and added CHANGELOG entry if required~
- [ ] ~Determined and documented upgrade steps~
- [ ] ~Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

## Testing Performed

Unit tests.

`curl -u admin:admin 'https://localhost:8000/api/product/usage/secured-units/csv?from=2023-08-17T15:36:24Z&to=2023-08-18T15:06:19Z' -ki`
```
HTTP/2 200 
content-disposition: attachment; filename="secured_units_usage-2023-08-17-2023-08-18.csv"
content-type: text/csv; charset=utf-8
vary: Accept-Encoding
content-length: 394
date: Wed, 23 Aug 2023 08:35:31 GMT

Timestamp,Nodes,CPU Units
2023-08-17T15:36:24Z,0,0
2023-08-17T16:12:02Z,0,0
2023-08-17T16:20:13Z,0,0
2023-08-17T17:20:13Z,1,8
2023-08-17T18:20:13Z,1,8
2023-08-17T19:20:13Z,1,8
2023-08-17T20:20:13Z,1,8
2023-08-17T21:20:13Z,1,8
2023-08-17T22:20:13Z,1,8
2023-08-18T10:29:21Z,1,8
2023-08-18T11:29:21Z,1,8
2023-08-18T12:29:21Z,1,8
2023-08-18T13:29:21Z,1,8
2023-08-18T14:06:19Z,0,0
```

Max:
`curl -u admin:admin 'https://localhost:8000/v1/product/usage/secured-units/max?from=2023-08-17T15:36:24Z&to=2023-08-18T15:06:19Z' -ki`
```
HTTP/2 200 
content-type: application/json
grpc-metadata-content-type: application/grpc
vary: Accept-Encoding
content-length: 129
date: Wed, 23 Aug 2023 08:36:47 GMT

{"maxNodesAt":"2023-08-17T17:20:13.282566375Z","maxNodes":"1","maxCpuUnitsAt":"2023-08-17T17:20:13.282566375Z","maxCpuUnits":"8"}
```

Current:
`curl -u admin:admin 'https://localhost:8000/v1/product/usage/secured-units/current' -ki`
```
HTTP/2 200 
content-type: application/json
grpc-metadata-content-type: application/grpc
vary: Accept-Encoding
content-length: 34
date: Wed, 23 Aug 2023 08:37:41 GMT

{"numNodes":"0","numCpuUnits":"0"}
```

### CODE REVIEW

File: central/productusage/datastore/securedunits/datastore_impl.go [hunk 1]
```diff
@@ -28,29 +32,80 @@ type dataStoreImpl struct {
...
-		if record.GetTimestamp().Compare(from) >= 0 && record.GetTimestamp().Compare(to) < 0 {
-			return fn(record)
+
+	pagination := search.NewPagination().
+		AddSortOption(search.NewSortOption(search.ProductUsageTimestamp)).Limit(page)
+
+	query := search.NewQueryBuilder().AddTimeRangeField(
+		search.ProductUsageTimestamp, from, to).WithPagination(pagination).ProtoQuery()
+
+	for offset := 0; ; offset += page {
```
Comment (connorgorman) [inline central/productusage/datastore/securedunits/datastore_impl.go:52]: This does one extra query than necessary. We should check that len(units) < page which means that we are on the last page. Only if the exact number of elements matches page then we'll do the extra query which should be rare

File: central/productusage/datastore/securedunits/datastore_impl.go [hunk 2]
```diff
@@ -28,29 +32,80 @@ type dataStoreImpl struct {
...
+func (ds *dataStoreImpl) GetMaxNumCPUUnits(ctx context.Context, from time.Time, to time.Time) (*storage.SecuredUnits, error) {
+	return ds.getMax(ctx, search.ProductUsageCPUUnits, from, to)
+}
+
+func (ds *dataStoreImpl) getMax(ctx context.Context, label search.FieldLabel, from time.Time, to time.Time) (*storage.SecuredUnits, error) {
+	if err := sac.VerifyAuthzOK(usageSAC.ReadAllowed(ctx)); err != nil {
+		return nil, errors.Wrap(err, "cannot permit to get maximum of usage data")
+	}
+	if from.IsZero() {
+		from = time.Unix(0, 0)
```
Comment (connorgorman) [inline central/productusage/datastore/securedunits/datastore_impl.go:85]: why is this necessary?

File: central/productusage/datastore/securedunits/datastore_impl.go [hunk 3]
```diff
@@ -28,29 +32,80 @@ type dataStoreImpl struct {
...
+	}
+	if to.IsZero() {
+		to = time.Now()
+	}
+
+	pagination := search.NewPagination().
```
Comment (connorgorman) [inline central/productusage/datastore/securedunits/datastore_impl.go:91]: The sort needs to be descending right? Default sort order is asc which is lowest value first

File: central/productusage/datastore/securedunits/datastore_impl_test.go [hunk 4]
```diff
@@ -78,34 +79,78 @@ func makeSource(n int64, c int64) *storage.SecuredUnits {
...
+	}
+
+	err = suite.datastore.Walk(suite.hasReadCtx, now, from, fn)
 	suite.NoError(err)
-	err = suite.datastore.Walk(suite.hasWriteCtx, &types.Timestamp{}, &types.Timestamp{}, fn)
+	suite.Equal(int64((N-1)*N/2), totalNodes)
+	suite.Equal(int64((N-1)*N), totalCPUUnits)
+}
+
+func (suite *UsageDataStoreTestSuite) TestGetMax() {
```
Comment (connorgorman) [inline central/productusage/datastore/securedunits/datastore_impl_test.go:123]: I would consider writing an integration test that uses a real postgres DB to ensure that the queries are working correctly

**Review by connorgorman — CHANGES_REQUESTED** [review]
Some comments about the implementation of finding the max that probably need to be corrected

**Comment by 0x656b694d** [comment]
Current dependencies on/for this PR:

* master
  * **PR #7403**
    * **PR #7136**
      * **PR #7137**
        * **PR #7138**
          * **PR #7179**
            * **PR #7472** 👈
          * **PR #7160**
          * **PR #7053**

**Review by connorgorman — APPROVED** [review]
Small comments on the error reporting, but lg, thanks for all the changes

**Pull request merged**